flink-runtime/src/main/java/org/apache/flink/streaming/api/functions/source/InputFormatSource.java

```diff
@@ -0,0 +1,314 @@
+    private final InputFormat<OUT, InputSplit> format;
+
+    @SuppressWarnings("unchecked")
+    public InputFormatSource(Boundedness boundedness, InputFormat<OUT, ?> format) {
+        this.boundedness = boundedness;
+        this.format = (InputFormat<OUT, InputSplit>) format;
+    }
+
+    @Override
+    public Boundedness getBoundedness() {
+        return boundedness;
+    }
+
+    @Override
+    public SplitEnumerator<SourceSplit, Void> createEnumerator(
+            SplitEnumeratorContext<SourceSplit> context) throws Exception {
+        return new InputFormatSplitEnumerator<>(format, context);
+    }
+
+    @Override
+    public SplitEnumerator<SourceSplit, Void> restoreEnumerator(
+            SplitEnumeratorContext<SourceSplit> context, Void checkpoint) throws Exception {
+        return new InputFormatSplitEnumerator<>(format, context);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<SourceSplit> getSplitSerializer() {
+        return new SimpleVersionedSerializer<>() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
+
+            @Override
+            public byte[] serialize(SourceSplit split) throws IOException {
+                return InstantiationUtil.serializeObject(split);
+            }
+
+            @Override
+            public SourceSplit deserialize(int version, byte[] serialized) throws IOException {
+                try {
+                    return InstantiationUtil.deserializeObject(
+                            serialized, SourceSplit.class.getClassLoader());
+                } catch (ClassNotFoundException e) {
+                    throw new IOException("Failed to deserialize SourceSplit.");
+                }
+            }
+        };
+    }
+
+    @Override
+    public SimpleVersionedSerializer<Void> getEnumeratorCheckpointSerializer() {
+        return new SimpleVersionedSerializer<>() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
+
+            @Override
+            public byte[] serialize(Void obj) {
+                return new byte[0];
+            }
+
+            @Override
+            public Void deserialize(int version, byte[] serialized) {
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public SourceReader<OUT, SourceSplit> createReader(SourceReaderContext readerContext)
+            throws Exception {
+        Preconditions.checkState(
+                readerContext instanceof RichSourceReaderContext,
+                "Source reader context must be an instance of RichSourceReaderContext");
+        RichSourceReaderContext richSourceReaderContext = (RichSourceReaderContext) readerContext;
+        RuntimeContext runtimeContext = richSourceReaderContext.getRuntimeContext();
+        Counter completedSplitsCounter =
+                runtimeContext.getMetricGroup().counter("numSplitsProcessed");
+
+        return new SourceReader<>() {
```

Comment:
Should this be an inner-class? That would make it easily unit testable if we find any issues later

Reply:
Moved.